nfs/src/rpc_task.cpp

```diff
@@ -318,6 +435,206 @@ void rpc_task::run_lookup()
     } while (rpc_retry);
 }
 
+std::vector<bytes_chunk>
+copy_to_cache(struct nfs_client *const client,
+        fuse_ino_t ino, struct fuse_bufvec* bufv, off_t offset, size_t length, int &error)
+{
+    // Check the valid ino.
+    assert(ino != 0);
+
+    const char* buf = (char *) bufv->buf[bufv->idx].mem  + bufv->off;
+
+    // Get file_inode structure from inode and chunk vec.
+    struct nfs_inode *inode = client->get_nfs_inode_from_ino(ino);
+    auto chunkvec = inode->filecache_handle->get(offset, length);
+    
+    // Result of pread/read.
+    size_t rem_len = length;
+
+    for (auto &chunk : chunkvec)
+    {
+        auto membuf = chunk.get_membuf();
+
+        // Lock the membuf to do the operation.
+        membuf->set_locked();
+
+        assert(rem_len >= chunk.length);
+
+        // Chunk is owned by us or uptodate.
+        if (chunk.is_empty || membuf->is_uptodate()) {
+            // buffer copy case.
+            memcpy(chunk.get_buffer(), buf, chunk.length);
+        } else {
+            // Need to issue read.
+            assert(0);
+        }
+
+        membuf->set_dirty();
+        membuf->set_uptodate();
+        membuf->clear_locked();
+
+        if (buf != nullptr) {
+            buf += chunk.length;
+        }
+
+        rem_len -= chunk.length;
+        assert(rem_len >= 0);
+    }
+
+    if (rem_len > 0) {
+        AZLogCrit("Remaining length not zero rem_len:{}", rem_len);
+
+        assert(0);
+    }
+    
+    return chunkvec;
+}
+
+/*
+ * sync_membuf do the following:-
+ * 1. It creates new flush_task.
+ * 2. Flush the membuf to backend asynchronously, while membuf is locked.
+ * 3. On completion write_flush_callback() called.
+ * 4. write_flush_callback() clear the dirty bit, release lock and release this chunk from chunkmap.
+ * 5. Free the flush_task.
+ *
+ * Note:- By creating new task, back pressure can be put on incoming writes. Otherwise
+ *        application keep issuing new writes and it can cause memory pressure on system.
+ *
+ * Note:- Write call don't need to wait for completion of this async task.
+ *        After starting async task, we can complete the write request, as write buffer
+ *        already copied to membuf.
+ *
+ * Note:- Flush can wait for this calls to finish by waiting on membuf lock.
+ *
+ */
+void sync_membuf(struct bytes_chunk& chunk, struct nfs_client *client, fuse_ino_t file_ino)
+{
+    WRITE3args args;
+    bool rpc_retry = false;
+    auto membuf = chunk.get_membuf();
+
+    // Verify the membuf.
+    assert(membuf != nullptr);
+    assert(membuf->is_inuse());
+
+    // Get the membuf which need to flush.
+    membuf->set_locked();
+    if (membuf->is_dirty() && !membuf->is_flushing())
```

Comment:
is_flushing check must be before the set_locked(), o/w how do we get the benefit of not having to wait for lock if someone is already fliushing.